pyretailscience/analysis/product_association.py

```diff
@@ -226,83 +204,162 @@ def _calc_association(  # noqa: C901 (ignore complexity) - Excluded due to min_*
+        )
 
-        occurrences = np.array(sparse_matrix.sum(axis=0)).flatten()
-        occurence_prob = occurrences / row_count
+        merged_df = ibis.join(
+            merged_df,
+            product_occurrences_1,
+            predicates=[merged_df["item_1"] == product_occurrences_1["item_1"]],
+        )
 
-        base_items = [target_item]
-        if number_of_combinations == 2:  # noqa: PLR2004
-            if target_item is None:
-                base_items = unique_combo_df[value_col].cat.categories
-            items = [([unique_combo_df[value_col].cat.categories.get_loc(cat)], cat) for cat in base_items]
-        elif number_of_combinations == 3:  # noqa: PLR2004
-            if target_item is None:
-                base_items = sorted(combinations(unique_combo_df[value_col].cat.categories, 2))
-            items = [
-                ([unique_combo_df[value_col].cat.categories.get_loc(i) for i in cats], cats) for cats in base_items
-            ]
+        merged_df = ibis.join(
+            merged_df,
+            product_occurrences_2,
+            predicates=[merged_df["item_2"] == product_occurrences_2["item_2"]],
+        )
 
-        if show_progress:
-            items = tqdm(items)
-
-        cols_mask = np.zeros(sparse_matrix.shape[1], dtype=bool)
-
-        for target_item_loc, item_2 in items:
-            target_item_col_index = cols_mask.copy()
-            target_item_col_index[target_item_loc] = True
-            rows_with_target_item = sparse_matrix[:, target_item_col_index].getnnz(axis=1) == len(target_item_loc)
-
-            cooccurrences = np.array(sparse_matrix[rows_with_target_item, :].sum(axis=0)).flatten()
-            if (cooccurrences == 0).all():
-                continue
-
-            rows_with_target_item_sum = rows_with_target_item.sum()
-            coocurrence_prob = cooccurrences / row_count
-
-            target_prob = rows_with_target_item_sum / row_count
-            expected_prob = target_prob * occurence_prob
-
-            # TODO: Try to avoid constructing a pandas Dataframe
-            pa_df = pd.DataFrame(
-                {
-                    f"{value_col}_1": [item_2] * sparse_matrix.shape[1],
-                    f"{value_col}_2": unique_combo_df[value_col].cat.categories.values,
-                    "occurrences_1": rows_with_target_item_sum,
-                    "occurrences_2": occurrences,
-                    "cooccurrences": cooccurrences,
-                    "support": coocurrence_prob,
-                    "confidence": cooccurrences / rows_with_target_item_sum,
-                    "uplift": coocurrence_prob / expected_prob,
-                },
-            )
+        cooccurrences = merged_df.group_by(["item_1", "item_2"]).aggregate(cooccurrences=merged_df[group_col].nunique())
+        cooccurrences = cooccurrences.mutate(
+            support=cooccurrences.cooccurrences / total_transactions,
```

Comment:
I think you can remove the trailing , here too
```suggestion
            support=cooccurrences.cooccurrences / total_transactions
```